File: app/src/main/java/com/krystianwsul/checkme/gui/customtimes/ShowCustomTimesFragment.java
package com.krystianwsul.checkme.gui.customtimes;


import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.gui.AbstractFragment;
import com.krystianwsul.checkme.gui.FabUser;
import com.krystianwsul.checkme.gui.SelectionCallback;
import com.krystianwsul.checkme.loaders.ShowCustomTimesLoader;
import com.krystianwsul.checkme.persistencemodel.SaveService;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.List;

public class ShowCustomTimesFragment extends AbstractFragment implements LoaderManager.LoaderCallbacks<ShowCustomTimesLoader.Data>, FabUser {
    private static final String SELECTED_CUSTOM_TIME_IDS_KEY = "selectedCustomTimeIds";

    private RecyclerView mShowTimesList;
    private CustomTimesAdapter mCustomTimesAdapter;
    private TextView mEmptyText;

    private ArrayList<Integer> mSelectedCustomTimeIds;

    private final SelectionCallback mSelectionCallback = new SelectionCallback() {
        @Override
        protected void unselect() {
            mCustomTimesAdapter.unselect();
        }

        @Override
        protected void onMenuClick(MenuItem menuItem) {
            ArrayList<Integer> customTimeIds = mCustomTimesAdapter.getSelected();
            Assert.assertTrue(customTimeIds != null);
            Assert.assertTrue(!customTimeIds.isEmpty());

            switch (menuItem.getItemId()) {
                case R.id.action_custom_times_delete:
                    mCustomTimesAdapter.removeSelected();

                    updateSelectAll();

                    break;
                default:
                    throw new UnsupportedOperationException();
            }
        }

        @Override
        protected void onFirstAdded() {
            ((AppCompatActivity) getActivity()).startSupportActionMode(this);

            mActionMode.getMenuInflater().inflate(R.menu.menu_custom_times, mActionMode.getMenu());

            updateFabVisibility();

            ((CustomTimesListListener) getActivity()).onCreateCustomTimesActionMode(mActionMode);
        }

        @Override
        protected void onSecondAdded() {

        }

        @Override
        protected void onOtherAdded() {

        }

        @Override
        protected void onLastRemoved() {
            updateFabVisibility();

            ((CustomTimesListListener) getActivity()).onDestroyCustomTimesActionMode();
        }

        @Override
        protected void onSecondToLastRemoved() {

        }

        @Override
        protected void onOtherRemoved() {

        }
    };

    @Nullable
    private FloatingActionButton mShowTimesFab;

    private ShowCustomTimesLoader.Data mData;

    public static ShowCustomTimesFragment newInstance() {
        return new ShowCustomTimesFragment();
    }

    public ShowCustomTimesFragment() {

    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        Assert.assertTrue(context instanceof CustomTimesListListener);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_show_custom_times, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        View view = getView();
        Assert.assertTrue(view != null);

        mShowTimesList = view.findViewById(R.id.show_times_list);
        mShowTimesList.setLayoutManager(new LinearLayoutManager(getActivity()));

        mEmptyText = view.findViewById(R.id.empty_text);
        Assert.assertTrue(mEmptyText != null);

        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(SELECTED_CUSTOM_TIME_IDS_KEY)) {
                mSelectedCustomTimeIds = savedInstanceState.getIntegerArrayList(SELECTED_CUSTOM_TIME_IDS_KEY);
                Assert.assertTrue(mSelectedCustomTimeIds != null);
                Assert.assertTrue(!mSelectedCustomTimeIds.isEmpty());
            }
        }

        getLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<ShowCustomTimesLoader.Data> onCreateLoader(int id, Bundle args) {
        return new ShowCustomTimesLoader(getActivity());
    }

    @Override
    public void onLoadFinished(Loader<ShowCustomTimesLoader.Data> loader, ShowCustomTimesLoader.Data data) {
        Assert.assertTrue(data != null);

        mData = data;

        if (mCustomTimesAdapter != null) {
            ArrayList<Integer> selectedCustomTimeIds = mCustomTimesAdapter.getSelected();
            if (selectedCustomTimeIds.isEmpty())
                mSelectedCustomTimeIds = null;
            else
                mSelectedCustomTimeIds = selectedCustomTimeIds;
        }

        mCustomTimesAdapter = new CustomTimesAdapter(data, this, mSelectedCustomTimeIds);
        mShowTimesList.setAdapter(mCustomTimesAdapter);

        mSelectionCallback.setSelected(mCustomTimesAdapter.getSelected().size());

        if (data.Entries.isEmpty()) {
            mShowTimesList.setVisibility(View.GONE);
            mEmptyText.setVisibility(View.VISIBLE);
            mEmptyText.setText(R.string.custom_times_empty);
        } else {
            mShowTimesList.setVisibility(View.VISIBLE);
            mEmptyText.setVisibility(View.GONE);
        }

        updateSelectAll();

        updateFabVisibility();
    }

    @Override
    public void onLoaderReset(Loader<ShowCustomTimesLoader.Data> loader) {
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (mCustomTimesAdapter != null) {
            ArrayList<Integer> selectedCustomTimeIds = mCustomTimesAdapter.getSelected();
            if (!selectedCustomTimeIds.isEmpty())
                outState.putIntegerArrayList(SELECTED_CUSTOM_TIME_IDS_KEY, selectedCustomTimeIds);
        }
    }

    private void updateSelectAll() {
        Assert.assertTrue(mCustomTimesAdapter != null);

        ((CustomTimesListListener) getActivity()).setCustomTimesSelectAllVisibility(mCustomTimesAdapter.getItemCount() != 0);
    }

    public void selectAll() {
        mCustomTimesAdapter.selectAll();
    }

    @Override
    public void setFab(@NonNull FloatingActionButton floatingActionButton) {
        mShowTimesFab = floatingActionButton;

        mShowTimesFab.setOnClickListener(v -> startActivity(ShowCustomTimeActivity.getCreateIntent(getActivity())));

        updateFabVisibility();
    }

    private void updateFabVisibility() {
        if (mShowTimesFab == null)
            return;

        if (mData != null && !mSelectionCallback.hasActionMode()) {
            mShowTimesFab.show();
        } else {
            mShowTimesFab.hide();
        }
    }

    @Override
    public void clearFab() {
        if (mShowTimesFab == null)
            return;

        mShowTimesFab.setOnClickListener(null);

        mShowTimesFab = null;
    }

    class CustomTimesAdapter extends RecyclerView.Adapter<CustomTimesAdapter.CustomTimeHolder> {
        private final int mDataId;
        private final List<CustomTimeWrapper> mCustomTimeWrappers;

        @NonNull
        private final ShowCustomTimesFragment mShowCustomTimesFragment;

        CustomTimesAdapter(@NonNull ShowCustomTimesLoader.Data data, @NonNull ShowCustomTimesFragment showCustomTimesFragment, @Nullable ArrayList<Integer> selectedCustomTimeIds) {
            mDataId = data.DataId;
            mCustomTimeWrappers = Stream.of(data.Entries)
                    .map(customTimeData -> new CustomTimeWrapper(customTimeData, selectedCustomTimeIds))
                    .collect(Collectors.toList());

            mShowCustomTimesFragment = showCustomTimesFragment;
        }

        @Override
        public int getItemCount() {
            return mCustomTimeWrappers.size();
        }

        void unselect() {
            Stream.of(mCustomTimeWrappers)
                    .filter(customTimeWrapper -> customTimeWrapper.mSelected)
                    .forEach(customTimeWrapper -> {
                        customTimeWrapper.mSelected = false;
                        notifyItemChanged(mCustomTimeWrappers.indexOf(customTimeWrapper));
                    });
        }

        public void selectAll() {
            Assert.assertTrue(!mSelectionCallback.hasActionMode());

            Stream.of(mCustomTimeWrappers)
                    .filter(customTimeWrapper -> !customTimeWrapper.mSelected)
                    .forEach(CustomTimeWrapper::toggleSelect);
        }

        @Override
        public CustomTimeHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(mShowCustomTimesFragment.getActivity());
            View showCustomTimesRow = layoutInflater.inflate(R.layout.row_show_custom_times, parent, false);

            TextView timesRowName = showCustomTimesRow.findViewById(R.id.times_row_name);

            return new CustomTimeHolder(showCustomTimesRow, timesRowName);
        }

        @Override
        public void onBindViewHolder(final CustomTimeHolder customTimeHolder, int position) {
            CustomTimeWrapper customTimeWrapper = mCustomTimeWrappers.get(position);
            Assert.assertTrue(customTimeWrapper != null);

            customTimeHolder.mTimesRowName.setText(customTimeWrapper.mCustomTimeData.Name);

            if (customTimeWrapper.mSelected)
                customTimeHolder.mShowCustomTimeRow.setBackgroundColor(ContextCompat.getColor(mShowCustomTimesFragment.getActivity(), R.color.selected));
            else
                customTimeHolder.mShowCustomTimeRow.setBackgroundColor(Color.TRANSPARENT);

            customTimeHolder.mShowCustomTimeRow.setOnLongClickListener(v -> {
                customTimeHolder.onLongClick();
                return true;
            });

            customTimeHolder.mShowCustomTimeRow.setOnClickListener(v -> {
                if (mShowCustomTimesFragment.mSelectionCallback.hasActionMode())
                    customTimeHolder.onLongClick();
                else
                    customTimeHolder.onRowClick();
            });
        }

        ArrayList<Integer> getSelected() {
            return new ArrayList<>(Stream.of(mCustomTimeWrappers)
                .filter(customTimeWrapper -> customTimeWrapper.mSelected)
                .map(customTimeWrapper -> customTimeWrapper.mCustomTimeData.Id)
                .collect(Collectors.toList()));
        }

        void removeSelected() {
            List<CustomTimeWrapper> selectedCustomTimeWrappers = Stream.of(mCustomTimeWrappers)
                    .filter(customTimeWrapper -> customTimeWrapper.mSelected)
                    .collect(Collectors.toList());

            for (CustomTimeWrapper customTimeWrapper : selectedCustomTimeWrappers) {
                int position = mCustomTimeWrappers.indexOf(customTimeWrapper);
                mCustomTimeWrappers.remove(position);
                notifyItemRemoved(position);
            }

            List<Integer> selectedCustomTimeIds = Stream.of(selectedCustomTimeWrappers)
                    .map(customTimeWrapper -> customTimeWrapper.mCustomTimeData.Id)
                    .collect(Collectors.toList());

            DomainFactory.getDomainFactory(mShowCustomTimesFragment.getActivity()).setCustomTimeCurrent(mShowCustomTimesFragment.getActivity(), mDataId, SaveService.Source.GUI, selectedCustomTimeIds);
        }

        class CustomTimeHolder extends RecyclerView.ViewHolder {
            final View mShowCustomTimeRow;
            final TextView mTimesRowName;

            CustomTimeHolder(View showCustomTimesRow, TextView timesRowName) {
                super(showCustomTimesRow);

                Assert.assertTrue(timesRowName != null);

                mShowCustomTimeRow = showCustomTimesRow;
                mTimesRowName = timesRowName;
            }

            void onRowClick() {
                CustomTimeWrapper customTimeWrapper = mCustomTimeWrappers.get(getAdapterPosition());
                Assert.assertTrue(customTimeWrapper != null);

                mShowCustomTimesFragment.getActivity().startActivity(ShowCustomTimeActivity.getEditIntent(customTimeWrapper.mCustomTimeData.Id, mShowCustomTimesFragment.getActivity()));
            }

            void onLongClick() {
                int position = getAdapterPosition();

                CustomTimeWrapper customTimeWrapper = mCustomTimeWrappers.get(position);
                Assert.assertTrue(customTimeWrapper != null);

                customTimeWrapper.toggleSelect();
            }
        }
    }

    private class CustomTimeWrapper {
        final ShowCustomTimesLoader.CustomTimeData mCustomTimeData;
        boolean mSelected = false;

        CustomTimeWrapper(ShowCustomTimesLoader.CustomTimeData customTimeData, ArrayList<Integer> selectedCustomTimeIds) {
            Assert.assertTrue(customTimeData != null);

            mCustomTimeData = customTimeData;

            if (selectedCustomTimeIds != null) {
                Assert.assertTrue(!selectedCustomTimeIds.isEmpty());

                mSelected = selectedCustomTimeIds.contains(mCustomTimeData.Id);
            }
        }

        void toggleSelect() {
            mSelected = !mSelected;

            if (mSelected) {
                mSelectionCallback.incrementSelected();
            } else {
                mSelectionCallback.decrementSelected();
            }

            int position = mCustomTimesAdapter.mCustomTimeWrappers.indexOf(this);
            Assert.assertTrue(position >= 0);

            mCustomTimesAdapter.notifyItemChanged(position);
        }
    }

    public interface CustomTimesListListener {
        void onCreateCustomTimesActionMode(ActionMode actionMode);
        void onDestroyCustomTimesActionMode();
        void setCustomTimesSelectAllVisibility(boolean selectAllVisible);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/customtimes/ShowCustomTimesFragment.kt
package com.krystianwsul.checkme.gui.customtimes


import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.CustomItemAnimator
import com.jakewharton.rxrelay3.BehaviorRelay
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.FragmentShowCustomTimesBinding
import com.krystianwsul.checkme.domainmodel.extensions.setCustomTimesCurrent
import com.krystianwsul.checkme.domainmodel.update.AndroidDomainUpdater
import com.krystianwsul.checkme.gui.base.AbstractFragment
import com.krystianwsul.checkme.gui.base.ActionModeListener
import com.krystianwsul.checkme.gui.base.SnackbarListener
import com.krystianwsul.checkme.gui.main.FabUser
import com.krystianwsul.checkme.gui.tree.AbstractHolder
import com.krystianwsul.checkme.gui.tree.AbstractModelNode
import com.krystianwsul.checkme.gui.tree.BaseAdapter
import com.krystianwsul.checkme.gui.tree.HolderType
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineDelegate
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineModelNode
import com.krystianwsul.checkme.gui.tree.delegates.multiline.MultiLineRow
import com.krystianwsul.checkme.gui.utils.BottomFabMenuDelegate
import com.krystianwsul.checkme.gui.utils.ResettableProperty
import com.krystianwsul.checkme.gui.utils.SelectionCallback
import com.krystianwsul.checkme.gui.widgets.MyBottomBar
import com.krystianwsul.checkme.utils.animateVisibility
import com.krystianwsul.checkme.viewmodels.ShowCustomTimesViewModel
import com.krystianwsul.checkme.viewmodels.getViewModel
import com.krystianwsul.common.utils.CustomTimeKey
import com.krystianwsul.treeadapter.*
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign
import java.util.*

class ShowCustomTimesFragment : AbstractFragment(), FabUser {

    companion object {

        private const val SELECTED_CUSTOM_TIME_IDS_KEY = "selectedCustomTimeIds"

        fun newInstance() = ShowCustomTimesFragment()
    }

    lateinit var treeViewAdapter: TreeViewAdapter<AbstractHolder>
        private set

    private var selectedCustomTimeKeys: List<CustomTimeKey>? = null

    private val listener get() = activity as CustomTimesListListener

    private val selectionCallback = object : SelectionCallback() {

        override val activity get() = requireActivity()

        override fun getTreeViewAdapter() = treeViewAdapter

        override fun unselect(placeholder: TreeViewAdapter.Placeholder) = treeViewAdapter.unselect(placeholder)

        override val bottomBarData by lazy { Triple(listener.getBottomBar(), R.menu.menu_custom_times, listener::initBottomBar) }

        override fun onMenuClick(itemId: Int, placeholder: TreeViewAdapter.Placeholder): Boolean {
            val customTimeIds = selectedIds
            check(customTimeIds.isNotEmpty())

            when (itemId) {
                R.id.action_custom_times_delete -> {
                    val selectedCustomTimeKeys =
                            (treeViewAdapter.treeModelAdapter as CustomTimesAdapter).customTimeWrappers
                                    .filter { it.treeNode.isSelected }
                                    .map { it.customTimeData.id }

                    fun setAreCurrent(current: Boolean) = AndroidDomainUpdater.setCustomTimesCurrent(
                            showCustomTimesViewModel.dataId.toFirst(),
                            selectedCustomTimeKeys,
                            current,
                    )

                    setAreCurrent(false).observeOn(AndroidSchedulers.mainThread())
                            .andThen(customTimesListListener.showSnackbarRemovedMaybe(selectedCustomTimeKeys.size))
                            .flatMapCompletable { setAreCurrent(true) }
                            .subscribe()
                            .addTo(createDisposable)
                }
                else -> throw UnsupportedOperationException()
            }

            return true
        }

        override fun onFirstAdded(placeholder: TreeViewAdapter.Placeholder, initial: Boolean) {
            (activity as AppCompatActivity).startSupportActionMode(this)

            updateFabVisibility()

            (activity as CustomTimesListListener).onCreateActionMode(actionMode!!)

            super.onFirstAdded(placeholder, initial)
        }

        override fun onLastRemoved(placeholder: TreeViewAdapter.Placeholder) {
            updateFabVisibility()

            (activity as CustomTimesListListener).onDestroyActionMode()
        }
    }

    private var showTimesFabDelegate: BottomFabMenuDelegate.FabDelegate? = null

    private lateinit var data: ShowCustomTimesViewModel.Data

    private val isVisible = BehaviorRelay.createDefault(false)
    private lateinit var showCustomTimesViewModel: ShowCustomTimesViewModel

    private val customTimesListListener get() = activity as CustomTimesListListener

    private val bindingProperty = ResettableProperty<FragmentShowCustomTimesBinding>()
    private var binding by bindingProperty

    override fun onAttach(context: Context) {
        super.onAttach(context)

        check(context is CustomTimesListListener)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) = FragmentShowCustomTimesBinding.inflate(inflater, container, false).also { binding = it }.root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState?.containsKey(SELECTED_CUSTOM_TIME_IDS_KEY) == true) {
            selectedCustomTimeKeys = savedInstanceState.getParcelableArrayList(SELECTED_CUSTOM_TIME_IDS_KEY)!!
            check(selectedCustomTimeKeys!!.isNotEmpty())
        }

        showCustomTimesViewModel = getViewModel<ShowCustomTimesViewModel>().apply {
            viewCreatedDisposable += isVisible.subscribe { if (it) start() else stop() }

            viewCreatedDisposable += data.subscribe { onLoadFinished(it) }
        }
    }

    private fun onLoadFinished(data: ShowCustomTimesViewModel.Data) {
        this.data = data

        if (this::treeViewAdapter.isInitialized) {
            selectedCustomTimeKeys = treeViewAdapter.selectedNodes
                    .asSequence()
                    .map { (it.modelNode as CustomTimeNode).customTimeData.id }
                    .toList()
                    .takeIf { it.isNotEmpty() }

            treeViewAdapter.updateDisplayedNodes {
                (treeViewAdapter.treeModelAdapter as CustomTimesAdapter).initialize()

                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, false)
            }
        } else {
            val customTimesAdapter = CustomTimesAdapter()
            customTimesAdapter.initialize()
            treeViewAdapter = customTimesAdapter.treeViewAdapter

            binding.showTimesList.apply {
                adapter = treeViewAdapter
                itemAnimator = CustomItemAnimator()
            }

            treeViewAdapter.updateDisplayedNodes {
                selectionCallback.setSelected(treeViewAdapter.selectedNodes.size, it, true)
            }
        }

        val (show, hide) = if (data.entries.isEmpty()) {
            binding.showCustomTimesEmptyTextInclude
                    .emptyText
                    .setText(R.string.custom_times_empty)

            binding.showCustomTimesEmptyTextInclude.emptyTextLayout to binding.showTimesList
        } else {
            binding.showTimesList to binding.showCustomTimesEmptyTextInclude.emptyTextLayout
        }

        animateVisibility(show, hide, immediate = data.immediate)

        updateSelectAll()
        updateFabVisibility()
    }

    private val selectedIds
        get() = treeViewAdapter.selectedNodes
                .asSequence()
                .map { (it.modelNode as CustomTimeNode).customTimeData.id }
                .toSet()

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        if (this::treeViewAdapter.isInitialized) {
            selectedIds.takeIf { it.isNotEmpty() }?.let {
                outState.putParcelableArrayList(SELECTED_CUSTOM_TIME_IDS_KEY, ArrayList(it))
            }
        }
    }

    private fun updateSelectAll() =
        (activity as CustomTimesListListener).setCustomTimesSelectAllVisibility(treeViewAdapter.displayedNodes.isNotEmpty())

    override fun setFab(fabDelegate: BottomFabMenuDelegate.FabDelegate) {
        showTimesFabDelegate = fabDelegate

        showTimesFabDelegate!!.setOnClickListener { startActivity(ShowCustomTimeActivity.getCreateIntent(requireActivity())) }

        updateFabVisibility()

        isVisible.accept(true)
    }

    private fun updateFabVisibility() {
        showTimesFabDelegate?.let {
            if (this::data.isInitialized && !selectionCallback.hasActionMode) it.show() else it.hide()
        }
    }

    override fun clearFab() {
        isVisible.accept(false)

        showTimesFabDelegate = null
    }

    override fun onDestroyView() {
        bindingProperty.reset()

        super.onDestroyView()
    }

    private inner class CustomTimesAdapter : BaseAdapter(), ActionModeCallback by selectionCallback {

        lateinit var customTimeWrappers: MutableList<CustomTimeNode>
            private set

        val treeViewAdapter = TreeViewAdapter(
                this,
                TreeViewAdapter.PaddingData(R.layout.row_group_list_fab_padding, R.id.paddingProgress),
        )

        override lateinit var treeNodeCollection: TreeNodeCollection<AbstractHolder>
            private set

        fun initialize() {
            treeNodeCollection = TreeNodeCollection(treeViewAdapter)
            treeViewAdapter.setTreeNodeCollection(treeNodeCollection)

            customTimeWrappers = data.entries
                    .asSequence()
                    .map { CustomTimeNode(it) }
                    .toMutableList()

            treeNodeCollection.nodes = customTimeWrappers.map { it.initialize(treeNodeCollection) }
        }
    }

    private inner class CustomTimeNode(val customTimeData: ShowCustomTimesViewModel.CustomTimeData) :
            AbstractModelNode(),
            MultiLineModelNode {

        public override lateinit var treeNode: TreeNode<AbstractHolder>
            private set

        override val holderType = HolderType.MULTILINE

        override val id = customTimeData.id

        fun initialize(treeNodeCollection: TreeNodeCollection<AbstractHolder>) = TreeNode(
            this,
            treeNodeCollection,
            selectedCustomTimeKeys?.contains(customTimeData.id) ?: false,
        ).also {
            treeNode = it
            it.setChildTreeNodes(listOf())
        }

        override val rowsDelegate = object : MultiLineModelNode.RowsDelegate {

            private val name = MultiLineRow.Visible(customTimeData.name)
            private val details = MultiLineRow.Visible(customTimeData.details, R.color.textSecondary)

            override fun getRows(isExpanded: Boolean, allChildren: List<TreeNode<*>>) = listOf(name, details)
        }

        override val isSelectable = true

        override val parentNode: ModelNode<AbstractHolder>? = null

        override val delegates by lazy { listOf(MultiLineDelegate(this)) }

        override val widthKey
            get() = MultiLineDelegate.WidthKey(
                    0,
                    false,
                    false,
                    false,
            )

        override fun onClick(holder: AbstractHolder) = startActivity(ShowCustomTimeActivity.getEditIntent(customTimeData.id, requireActivity()))

        private val customTimeId get() = customTimeData.id.customTimeId.value

        override fun compareTo(other: ModelNode<AbstractHolder>) =
                customTimeId.compareTo((other as CustomTimeNode).customTimeId)
    }

    interface CustomTimesListListener : ActionModeListener, SnackbarListener {

        fun setCustomTimesSelectAllVisibility(selectAllVisible: Boolean)

        fun getBottomBar(): MyBottomBar

        fun initBottomBar()
    }
}
